src/cargo/core/compiler/job_queue.rs

```diff
@@ -251,26 +366,209 @@ impl<'a, 'cfg> JobQueue<'a, 'cfg> {
             .take()
             .map(move |srv| srv.start(move |msg| drop(tx.send(Message::FixDiagnostic(msg)))));
 
-        // Use `crossbeam` to create a scope in which we can execute scoped
-        // threads. Note that this isn't currently required by Cargo but it was
-        // historically required. This is left in for now in case we need the
-        // `'a` ability for child threads in the near future, but if this
-        // comment has been sitting here for a long time feel free to refactor
-        // away crossbeam.
-        crossbeam_utils::thread::scope(|scope| self.drain_the_queue(cx, plan, scope, &helper))
-            .expect("child threads shouldn't panic")
+        self.drain_the_queue(cx, plan, &helper)
+    }
+
+    fn spawn_work_if_possible(
+        &mut self,
+        cx: &mut Context<'a, '_>,
+        jobserver_helper: &HelperThread,
+        has_errored: bool,
+    ) -> CargoResult<()> {
+        // Dequeue as much work as we can, learning about everything
+        // possible that can run. Note that this is also the point where we
+        // start requesting job tokens. Each job after the first needs to
+        // request a token.
+        while let Some((unit, job)) = self.queue.dequeue() {
+            self.pending_queue.push((unit, job));
+        }
+
+        // Do not actually spawn the new work if we've errored out
+        if has_errored {
+            return Ok(());
+        }
+
+        // Arguably, this is overdoing it (by far!) -- we may have already
+        // requested many of these tokens.
+        //
+        // The jobserver crate would ideally expose an interface of "how many
+        // tokens I currently want" rather than just requesting a token (and
+        // providing no way to cancel said request). Pretty much every user of
+        // the jobserver crate likely wants to do something akin to that.
+        let tokens_needed = self
+            .unused_tokens
+            .len()
+            .saturating_sub(self.pending_queue.len());
+        for _ in 0..tokens_needed {
+            jobserver_helper.request_token();
```

Comment:
This I'm a little worried about since it will request a huge amount more than before. I don't think we want to do these requests on each turn of the loop, we should still strive to match up a request to an acquire 1:1